### PR TITLE
Fix TestRunner compatibility to PhpUnit 8

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no     
| Deprecations? | no
| Tests pass?   | yes   
| Fixed tickets | related to: #30055
| License       | MIT
| Doc PR        | -

Modify the installed phpunit version to be compatibility with the symfony custom TestRunner. This is sure not the best way but maybe currently the fastest way to support PhpUnit 8. The hack should be removed as soon as there is another way to implement a custom Runner.

### CODE REVIEW

File: src/Symfony/Bridge/PhpUnit/bin/simple-phpunit [hunk 1]
```diff
@@ -90,6 +90,18 @@ if (!file_exists("$PHPUNIT_DIR/phpunit-$PHPUNIT_VERSION/phpunit") || md5_file(__
     if ($exit) {
         exit($exit);
     }
+    if ($PHPUNIT_VERSION >= 8.0) {
+        // Fix compatibility to custom TestRunner for PhpUnit 8
+        $testRunnerFile = 'src/TextUI/TestRunner.php';
+        file_put_contents(
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/bin/simple-phpunit]: should be on one line

**Review by nicolas-grekas — COMMENTED** [review]
Before merging, we should be sure there is really no alternatives - we should also report the issue to Sebastian so that he knows some extensibility point is now missing. Could you please have a look?

**Comment by kunicmarko20** [comment]
Isn't it a better solution to check phpunit version here:

https://github.com/symfony/symfony/blob/e8c3f9e0d00ddd6cf72340c5f61fb78932a23e4c/src/Symfony/Bridge/PhpUnit/TextUI/TestRunner.php#L22-L26

and then provide class with or without final?

**Comment by nicolas-grekas** [comment]
@kunicmarko20 could you please have a look if that would be possible? Can you find any other alternative?

**Comment by alexander-schranz** [comment]
@nicolas-grekas is it maybe possible to remove the TestRunner and move this logic to the Test Command?

**Comment by fabpot** [comment]
@nicolas-grekas @alexander-schranz How can we move forward here?

**Comment by alexander-schranz** [comment]
Will have a look at it, at the _EU-FOSSA Hackathon_.

**Comment by ostrolucky** [comment]
Could this help? https://github.com/dg/bypass-finals

**Comment by alexander-schranz** [comment]
@ostrolucky think we don't want to remove all finals. 

Maybe there is some suggestion by @sebastianbergmann how we could automatically register the SymfonyListener programmatically now. Did create an issue at the phpunit repository https://github.com/sebastianbergmann/phpunit/issues/3594

#EU-FOSSA

**Comment by sebastianbergmann** [comment]
As I already wrote in https://github.com/sebastianbergmann/phpunit/issues/3594#issuecomment-480487689 (why do we have to discuss the same issue in different places?), "automatically registereing" is not possible, and intentionally so: the test runner should not be extended through inheritance but by hooking into certain events. You can find the available hooks [here](https://github.com/sebastianbergmann/phpunit/tree/master/src/Runner/Hook).

What you should do is implement a class that implements one or more of these `Hook` interfaces. Then have PHPUnit load that class through its configuration file. PHPUnit will then register the hooks.

I can only advise against "hacking around" or "extending" the PHPUnit test runner and then, again and again, update your workarounds when internal implementation details of PHPUnit change. Please do not rely on private implementation details.

In https://github.com/symfony/symfony/pull/30085#pullrequestreview-200951632 @nicolas-grekas wrote

> we should also report the issue to Sebastian so that he knows some extensibility point is now missing

Yes, please: let me know if you are missing an extension point. While working on https://github.com/Roave/no-leaks @Ocramius ran into limitations of the current hooks provided and we plan to work on this in the future.

On a related note, the whole idea of phpunit-bridge is a mystery to me. I do not know what problem(s) you are trying to solve with it. It would be great if somebody could explain that at some point to me. It might be that there are solutions in there for problems that could be solved differently, either inside PHPUnit or in Symfony. I am not conviced that the best medium for this explanation would be a ticket or email. This should, ideally, be a real conversation. Maybe somebody from Symfony can come to a PHPUnit code sprint (there is one in Würzburg this month) or I can come to a Symfony code sprint.

**Comment by nicolas-grekas** [comment]
Meeting in person would be great for sure! I'm pretty busy this spring so not sure I can join the next sprint but let's tell each other if we see about any opportunity!

The bridge creates at alternative "phpunit binary" that provides some extra features out of the box. With inheritance, it was possible to wire a different behavior by default, but now I'm not sure we can (I didn't have a deeper look). Which features? some extra annotations (`@expectedDeprecation`, `@time-sensitive`, etc) and some extra behavior: the deprecations report that is displayed after running the tests, or a record/replay feature for skipped tests.

The extra features related to annotations work with a line in config files, but the extra behaviors are shipped out of the box when using the bridge.

That's why there is this specific bootstrapping extra layer, which is closed now, with `final`.

**Comment by alexander-schranz** [comment]
Analysed more the current state and I think I could fix it by moving the logic from the TestRunner to the command.

#EUFOSSA

File: src/Symfony/Bridge/PhpUnit/Legacy/CommandForV6.php [hunk 2]
```diff
@@ -27,6 +27,24 @@ class CommandForV6 extends BaseCommand
      */
     protected function createRunner(): BaseRunner
     {
-        return new TestRunner($this->arguments['loader']);
+        $listener = new SymfonyTestsListener();
+
+        $this->arguments['listeners'] = isset($arguments['listeners']) ? $this->arguments['listeners'] : [];
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/Legacy/CommandForV6.php]: array() (php5.3)

Comment (nicolas-grekas) [reply src/Symfony/Bridge/PhpUnit/Legacy/CommandForV6.php]: oh, ok, no, `V6`

**Comment by nicolas-grekas** [comment]
if it works, perfect :)
wasn't `TestRunnerForV6` reference somewhere?

**Comment by xabbuh** [comment]
Can we also remove the `src/Symfony/Bridge/PhpUnit/TextUI/TestRunner.php` file?

**Comment by derrabus** [comment]
@xabbuh It's still used in the `CommandForV5`, isn't it?

**Comment by alexander-schranz** [comment]
@xabbuh done and move the v5 testrunner logic also to the commandforv5

File: src/Symfony/Bridge/PhpUnit/Legacy/CommandForV5.php [hunk 3]
```diff
@@ -23,6 +23,24 @@ class CommandForV5 extends \PHPUnit_TextUI_Command
      */
     protected function createRunner()
     {
-        return new TestRunnerForV5($this->arguments['loader']);
+        $listener = new SymfonyTestsListenerForV5();
+
+        $this->arguments['listeners'] = isset($this->arguments['listeners']) ? $this->arguments['listeners'] : array();
```
Comment (alexander-schranz) [inline src/Symfony/Bridge/PhpUnit/Legacy/CommandForV5.php]: @nicolas-grekas but here the long array syntax is correct or not?

**Review by nicolas-grekas — APPROVED** [review]
cool, thanks for digging!

**Comment by nicolas-grekas** [comment]
Thank you @alexander-schranz.

**Pull request merged**